File: Algorithms/p116_Populating_Next_Right_Pointers_in_Each_Node_2.cpp
/**
 * Definition for binary tree with next pointer.
 * struct TreeLinkNode {
 *  int val;
 *  TreeLinkNode *left, *right, *next;
 *  TreeLinkNode(int x) : val(x), left(NULL), right(NULL), next(NULL) {}
 * };
 */
class Solution {
public:
    void connect(TreeLinkNode *root) {
        if(!root) return;
        for(TreeLinkNode* left_most=root;left_most->left;left_most=left_most->left){
            for(TreeLinkNode* p=left_most;;p=p->next){
                p->left->next = p->right;
                if(!p->next) break;
                p->right->next = p->next->left;
            }
        }
    }
};

File: Algorithms/p116_Populating_Next_Right_Pointers_in_Each_Node_2.py
# Definition for binary tree with next pointer.
# class TreeLinkNode:
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None
#         self.next = None

class Solution:
    # @param root, a tree link node
    # @return nothing
    def connect(self, root):
        if not root:
            return 
        left_most = root
        while left_most.left:
            p = left_most
            while True:
                p.left.next = p.right
                if not p.next:
                    break
                p.right.next = p.next.left
                p = p.next
            left_most = left_most.left
